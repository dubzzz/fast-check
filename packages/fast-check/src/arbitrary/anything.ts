import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { anyArbitraryBuilder } from './_internals/builders/AnyArbitraryBuilder';
import type { ObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';
import { toQualifiedObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';

export type { ObjectConstraints };

/**
 * For any type of values
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(): Arbitrary<unknown>;
/**
 * For any type of values following the constraints defined by `settings`
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @example
 * ```typescript
 * // Using custom settings
 * fc.anything({
 *     key: fc.char(),
 *     values: [fc.integer(10,20), fc.constant(42)],
 *     maxDepth: 2
 * });
 * // Can build entries such as:
 * // - 19
 * // - [{"2":12,"k":15,"A":42}]
 * // - {"4":[19,13,14,14,42,11,20,11],"6":42,"7":16,"L":10,"'":[20,11],"e":[42,20,42,14,13,17]}
 * // - [42,42,42]...
 * ```
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(constraints: ObjectConstraints): Arbitrary<unknown>;
function anything(constraints?: ObjectConstraints): Arbitrary<unknown> {
  return anyArbitraryBuilder(toQualifiedObjectConstraints(constraints));
}
export { anything };
