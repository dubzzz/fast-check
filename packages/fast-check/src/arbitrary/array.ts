import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { ArrayArbitrary } from './_internals/ArrayArbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import {
  MaxLengthUpperBound,
  maxGeneratedLengthFromSizeForArbitrary,
} from './_internals/helpers/MaxLengthFromMinLength';
import type { DepthIdentifier } from './_internals/helpers/DepthContext';

/**
 * Constraints to be applied on {@link array}
 * @remarks Since 2.4.0
 * @public
 */
export interface ArrayConstraints {
  /**
   * Lower bound of the generated array size
   * @defaultValue 0
   * @remarks Since 2.4.0
   */
  minLength?: number;
  /**
   * Upper bound of the generated array size
   * @defaultValue 0x7fffffff â€” _defaulting seen as "max non specified" when `defaultSizeToMaxWhenMaxSpecified=true`_
   * @remarks Since 2.4.0
   */
  maxLength?: number;
  /**
   * Define how large the generated values should be (at max)
   *
   * When used in conjonction with `maxLength`, `size` will be used to define
   * the upper bound of the generated array size while `maxLength` will be used
   * to define and document the general maximal length allowed for this case.
   *
   * @remarks Since 2.22.0
   */
  size?: SizeForArbitrary;
  /**
   * When receiving a depth identifier, the arbitrary will impact the depth
   * attached to it to avoid going too deep if it already generated lots of items.
   *
   * In other words, if the number of generated values within the collection is large
   * then the generated items will tend to be less deep to avoid creating structures a lot
   * larger than expected.
   *
   * For the moment, the depth is not taken into account to compute the number of items to
   * define for a precise generate call of the array. Just applied onto eligible items.
   *
   * @remarks Since 2.25.0
   */
  depthIdentifier?: DepthIdentifier | string;
}

/**
 * Extra but internal constraints that can be passed to array.
 * This extra set is made of constraints mostly targets experimental and internal features not yet mature to be exposed.
 * @internal
 */
export interface ArrayConstraintsInternal<T> extends ArrayConstraints {
  /**
   * Extra user-definable and hardcoded values.
   * Each entry in the array could be used to build the final generated value outputed by the arbitrary of array on generate.
   * Each entry must have at least one element of type T into it.
   * Each T must be a value acceptable for the arbitrary passed to the array.
   */
  experimentalCustomSlices?: T[][];
}

/**
 * For arrays of values coming from `arb`
 *
 * @param arb - Arbitrary used to generate the values inside the array
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function array<T>(arb: Arbitrary<T>, constraints: ArrayConstraints = {}): Arbitrary<T[]> {
  const size = constraints.size;
  const minLength = constraints.minLength || 0;
  const maxLengthOrUnset = constraints.maxLength;
  const depthIdentifier = constraints.depthIdentifier;
  const maxLength = maxLengthOrUnset !== undefined ? maxLengthOrUnset : MaxLengthUpperBound;
  const specifiedMaxLength = maxLengthOrUnset !== undefined;
  const maxGeneratedLength = maxGeneratedLengthFromSizeForArbitrary(size, minLength, maxLength, specifiedMaxLength);
  const customSlices = (constraints as ArrayConstraintsInternal<T>).experimentalCustomSlices || [];
  return new ArrayArbitrary<T>(arb, minLength, maxGeneratedLength, maxLength, depthIdentifier, undefined, customSlices);
}
export { array };
