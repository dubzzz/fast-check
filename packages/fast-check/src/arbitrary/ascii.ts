import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';
import { indexToPrintableIndexMapper, indexToPrintableIndexUnmapper } from './_internals/mappers/IndexToPrintableIndex';

/**
 * For single ascii characters - char code between 0x00 (included) and 0x7f (included)
 * @remarks Since 0.0.1
 * @public
 */
export function ascii(): Arbitrary<string> {
  return buildCharacterArbitrary(0x00, 0x7f, indexToPrintableIndexMapper, indexToPrintableIndexUnmapper);
}
