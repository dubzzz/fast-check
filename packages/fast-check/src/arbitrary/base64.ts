import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';

/** @internal */
function base64Mapper(v: number) {
  if (v < 26) return v + 65; // A-Z
  if (v < 52) return v + 97 - 26; // a-z
  if (v < 62) return v + 48 - 52; // 0-9
  return v === 62 ? 43 : 47; // +/
}

/** @internal */
function base64Unmapper(v: number) {
  if (v >= 65 && v <= 90) return v - 65; // A-Z
  if (v >= 97 && v <= 122) return v - 97 + 26; // a-z
  if (v >= 48 && v <= 57) return v - 48 + 52; // 0-9
  return v === 43 ? 62 : v === 47 ? 63 : -1; // +/
}

/**
 * For single base64 characters - A-Z, a-z, 0-9, + or /
 * @remarks Since 0.0.1
 * @public
 */
export function base64(): Arbitrary<string> {
  return buildCharacterArbitrary(0, 63, base64Mapper, base64Unmapper);
}
