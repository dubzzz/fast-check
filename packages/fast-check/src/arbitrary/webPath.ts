import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';
import { resolveSize } from './_internals/helpers/MaxLengthFromMinLength';
import { buildUriPathArbitrary } from './_internals/builders/UriPathArbitraryBuilder';

/**
 * Constraints to be applied on {@link webPath}
 * @remarks Since 3.3.0
 * @public
 */
export interface WebPathConstraints {
  /**
   * Define how large the generated values should be (at max)
   * @remarks Since 3.3.0
   */
  size?: Exclude<SizeForArbitrary, 'max'>;
}

/**
 * For web path
 *
 * According to {@link https://www.ietf.org/rfc/rfc3986.txt | RFC 3986} and
 * {@link https://url.spec.whatwg.org/ | WHATWG URL Standard}
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 3.3.0
 * @public
 */
export function webPath(constraints?: WebPathConstraints): Arbitrary<string> {
  const c = constraints || {};
  const resolvedSize = resolveSize(c.size);
  return buildUriPathArbitrary(resolvedSize);
}
