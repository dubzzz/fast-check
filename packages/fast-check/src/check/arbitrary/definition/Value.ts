import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
