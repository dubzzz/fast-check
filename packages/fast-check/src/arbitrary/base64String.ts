import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { base64 } from './base64';
import { MaxLengthUpperBound } from './_internals/helpers/MaxLengthFromMinLength';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
import { stringToBase64Mapper, stringToBase64Unmapper } from './_internals/mappers/StringToBase64';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For base64 strings
 *
 * A base64 string will always have a length multiple of 4 (padded with =)
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
function base64String(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const { minLength: unscaledMinLength = 0, maxLength: unscaledMaxLength = MaxLengthUpperBound, size } = constraints;

  const minLength = unscaledMinLength + 3 - ((unscaledMinLength + 3) % 4);
  const maxLength = unscaledMaxLength - (unscaledMaxLength % 4);
  const requestedSize = constraints.maxLength === undefined && size === undefined ? '=' : size;

  if (minLength > maxLength) throw new Error('Minimal length should be inferior or equal to maximal length');
  if (minLength % 4 !== 0) throw new Error('Minimal length of base64 strings must be a multiple of 4');
  if (maxLength % 4 !== 0) throw new Error('Maximal length of base64 strings must be a multiple of 4');

  const charArbitrary = base64();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, codePointsToStringUnmapper);
  const enrichedConstraints: ArrayConstraintsInternal<string> = {
    minLength,
    maxLength,
    size: requestedSize,
    experimentalCustomSlices,
  };
  return array(charArbitrary, enrichedConstraints)
    .map(codePointsToStringMapper, codePointsToStringUnmapper)
    .map(stringToBase64Mapper, stringToBase64Unmapper);
}
export { base64String };
