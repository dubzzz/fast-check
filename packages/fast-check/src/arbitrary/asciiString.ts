import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { array } from './array';
import { ascii } from './ascii';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link ascii}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function asciiString(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return array(ascii(), constraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
