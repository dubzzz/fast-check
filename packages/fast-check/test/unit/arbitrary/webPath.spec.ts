import fc from 'fast-check';
import type { WebPathConstraints } from '../../../src/arbitrary/webPath';
import { webPath } from '../../../src/arbitrary/webPath';
import { URL } from 'url';

import {
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
} from './__test-helpers__/ArbitraryAssertions';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { relativeSizeArb, sizeArb } from './__test-helpers__/SizeHelpers';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('webPath (integration)', () => {
  type Extra = WebPathConstraints;

  const extraParametersBuilder = webPathConstraintsBuilder;

  const isCorrect = (path: string) => {
    // Valid path given the specs defined by WHATWG URL Standard: https://url.spec.whatwg.org/
    // A TypeError will be thrown if the input is not a valid URL: https://nodejs.org/api/url.html#url_constructor_new_url_input_base
    expect(() => new URL('http://domain' + path)).not.toThrow();
    if (path.length !== 0) {
      expect(path[0]).toBe('/');
    }
  };

  const webPathBuilder = (extra: Extra) => webPath(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(webPathBuilder, { extraParameters: extraParametersBuilder() });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(webPathBuilder, isCorrect, { extraParameters: extraParametersBuilder() });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(webPathBuilder, { extraParameters: extraParametersBuilder(true) });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(webPathBuilder, {
      extraParameters: extraParametersBuilder(true),
    });
  });

  it.each`
    rawValue
    ${'/a/z'}
    ${'/azerty'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = webPath();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});

// Helpers

function webPathConstraintsBuilder(onlySmall?: boolean): fc.Arbitrary<WebPathConstraints> {
  return fc.record(
    { size: onlySmall ? fc.constantFrom('-1', '=', 'xsmall', 'small') : fc.oneof(sizeArb, relativeSizeArb) },
    { requiredKeys: [] },
  );
}
