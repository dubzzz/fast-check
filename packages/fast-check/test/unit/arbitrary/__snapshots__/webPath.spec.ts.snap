// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webPath (integration) should be able to shrink /a/z 1`] = `
""/a/z"
├> ""
├> "/z"
|  ├> ""
|  ├> "/"
|  |  └> ""
|  ├> "/a"
|  |  ├> ""
|  |  └> "/"
|  |     └> ""
|  ├> "/n"
|  |  ├> ""
|  |  ├> "/"
|  |  |  └> ""
|  |  ├> "/h"
|  |  |  ├> ""
|  |  |  ├> "/"
|  |  |  |  └> ""
|  |  |  ├> "/e"
|  |  |  |  ├> ""
|  |  |  |  ├> "/"
|  |  |  |  |  └> ""
|  |  |  |  ├> "/c"
|  |  |  |  |  ├> ""
|  |  |  |  |  ├> "/"
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "/b"
|  |  |  |  |     ├> ""
|  |  |  |  |     ├> "/"
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> "/a"
|  |  |  |  |        ├> ""
|  |  |  |  |        └> "/"
|  |  |  |  |           └> ""
|  |  |  |  └> "/d"
|  |  |  |     ├> ""
|  |  |  |     ├> "/"
|  |  |  |     |  └> ""
|  |  |  |     └> "/c"
|  |  |  |        ├> ""
|  |  |  |        ├> "/"
|  |  |  |        |  └> ""
|  |  |  |        ├> "/a"
|  |  |  |        |  ├> ""
|  |  |  |        |  └> "/"
|  |  |  |        |     └> ""
|  |  |  |        └> "/b"
|  |  |  |           ├> ""
|  |  |  |           ├> "/"
|  |  |  |           |  └> ""
|  |  |  |           └> "/a"
|  |  |  |              ├> ""
|  |  |  |              └> "/"
|  |  |  |                 └> ""
|  |  |  └> "/g"
|  |  |     ├> ""
|  |  |     ├> "/"
|  |  |     |  └> ""
|  |  |     └> "/f"
|  |  |        ├> ""
|  |  |        ├> "/"
|  |  |        |  └> ""
|  |  |        └> "/e"
|  |  |           ├> ""
|  |  |           ├> "/"
|  |  |           |  └> ""
|  |  |           ├> "/a"
|  |  |           |  ├> ""
|  |  |           |  └> "/"
|  |  |           |     └> ""
|  |  |           ├> "/c"
|  |  |           |  ├> ""
|  |  |           |  ├> "/"
|  |  |           |  |  └> ""
|  |  |           |  └> "/b"
|  |  |           |     ├> ""
|  |  |           |     ├> "/"
|  |  |           |     |  └> ""
|  |  |           |     └> "/a"
|  |  |           |        ├> ""
|  |  |           |        └> "/"
|  |  |           |           └> ""
|  |  |           └> "/d"
|  |  |              ├> ""
|  |  |              ├> "/"
|  |  |              |  └> ""
|  |  |              └> "/c"
|  |  |                 ├> ""
|  |  |                 ├> "/"
|  |  |                 |  └> ""
|  |  |                 ├> "/a"
|  |  |                 |  ├> ""
|  |  |                 |  └> "/"
|  |  |                 |     └> ""
|  |  |                 └> "/b"
|  |  |                    ├> ""
|  |  |                    ├> "/"
|  |  |                    |  └> ""
|  |  |                    └> "/a"
|  |  |                       ├> ""
|  |  |                       └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`webPath (integration) should be able to shrink /azerty 1`] = `
""/azerty"
├> ""
├> "/"
|  └> ""
├> "/rty"
|  ├> ""
|  ├> "/ty"
|  |  ├> ""
|  |  ├> "/y"
|  |  |  ├> ""
|  |  |  ├> "/"
|  |  |  |  └> ""
|  |  |  ├> "/a"
|  |  |  |  ├> ""
|  |  |  |  └> "/"
|  |  |  |     └> ""
|  |  |  ├> "/m"
|  |  |  |  ├> ""
|  |  |  |  ├> "/"
|  |  |  |  |  └> ""
|  |  |  |  ├> "/g"
|  |  |  |  |  ├> ""
|  |  |  |  |  ├> "/"
|  |  |  |  |  |  └> ""
|  |  |  |  |  ├> "/d"
|  |  |  |  |  |  ├> ""
|  |  |  |  |  |  ├> "/"
|  |  |  |  |  |  |  └> ""
|  |  |  |  |  |  └> "/c"
|  |  |  |  |  |     ├> ""
|  |  |  |  |  |     ├> "/"
|  |  |  |  |  |     |  └> ""
|  |  |  |  |  |     └> "/b"
|  |  |  |  |  |        ├> ""
|  |  |  |  |  |        ├> "/"
|  |  |  |  |  |        |  └> ""
|  |  |  |  |  |        └> "/a"
|  |  |  |  |  |           ├> ""
|  |  |  |  |  |           └> "/"
|  |  |  |  |  |              └> ""
|  |  |  |  |  └> "/f"
|  |  |  |  |     ├> ""
|  |  |  |  |     ├> "/"
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> "/e"
|  |  |  |  |        ├> ""
|  |  |  |  |        ├> "/"
|  |  |  |  |        |  └> ""
|  |  |  |  |        └> "/d"
|  |  |  |  |           ├> ""
|  |  |  |  |           ├> "/"
|  |  |  |  |           |  └> ""
|  |  |  |  |           ├> "/a"
|  |  |  |  |           |  ├> ""
|  |  |  |  |           |  └> "/"
|  |  |  |  |           |     └> ""
|  |  |  |  |           └> "/c"
|  |  |  |  |              ├> ""
|  |  |  |  |              ├> "/"
|  |  |  |  |              |  └> ""
|  |  |  |  |              └> "/b"
|  |  |  |  |                 ├> ""
|  |  |  |  |                 ├> "/"
|  |  |  |  |                 |  └> ""
|  |  |  |  |                 └> "/a"
|  |  |  |  |                    ├> ""
|  |  |  |  |                    └> "/"
|  |  |  |  |                       └> ""
|  |  |  |  ├> "/j"
|  |  |  |  |  ├> ""
|  |  |  |  |  ├> "/"
|  |  |  |  |  |  └> ""
|  |  |  |  |  └> "/i"
|  |  |  |  |     ├> ""
|  |  |  |  |     ├> "/"
|  |  |  |  |     |  └> ""
|  |  |  |  |     └> "/h"
|  |  |  |  |        ├> ""
|  |  |  |  |        ├> "/"
|  |  |  |  |        |  └> ""
|  |  |  |  |        └> "/g"
|  |  |  |  |           ├> ""
|  |  |  |  |           ├> "/"
|  |  |  |  |           |  └> ""
|  |  |  |  |           ├> "/a"
|  |  |  |  |           |  ├> ""
|  |  |  |  |           |  └> "/"
|  |  |  |  |           |     └> ""
|  |  |  |  |           ├> "/d"
|  |  |  |  |           |  ├> ""
|  |  |  |  |           |  ├> "/"
|  |  |  |  |           |  |  └> ""
|  |  |  |  |           |  └> "/c"
|  |  |  |  |           |     ├> ""
|  |  |  |  |           |     ├> "/"
|  |  |  |  |           |     |  └> ""
|  |  |  |  |           |     └> "/b"
|  |  |  |  |           |        ├> ""
|  |  |  |  |           |        ├> "/"
|  |  |  |  |           |        |  └> ""
|  |  |  |  |           |        └> …
|  |  |  |  |           └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
